text/3735-field-projections.md

```diff
@@ -0,0 +1,1421 @@
+    }
+}
+
+unsafe impl<F1, F2> PinField for field_of!(FairRaceFuture<F1, F2>, fut2) {
+    type Projected<'a> = &'a mut bool;
+
+    fn from_pinned_ref<'a>(r: &'a mut bool) -> &'a mut bool {
+      r
+    }
+}
+```
+
+Now the only component that is left is an implementation of `Projectable` and `Project` for
+`Pin<&mut T>`:
+
+```rust
+impl<'a, T: ?Sized> Projectable for Pin<&'a mut T> {
+    type Inner = T;
+}
+
+impl<'a, T, F> Project<F> for Pin<&'a mut T>
+where
+    F: Field<Base = T> + PinField,
+{
+    type Output = <F as PinField>::Projected<'a>;
+
+    fn project(self) -> Self::Output {
+        let r: &mut T = unsafe { Pin::into_inner_unchecked(self) };
+        let r: &mut F::Type = <&mut T as Project::<F>>::project(r);
+        <F as PinField>::from_pinned_ref(r)
+    }
+}
+```
+
+## Interactions
+
+There aren't a lot of interactions with other features.
+
+The projection operator binds very tightly:
+
+```rust
+*ctr->field = *(ctr->field);
+
+&mut ctr->field = &mut (ctr->field);
+
+ctr->field.foo() = (ctr->field).foo();
+
+ctr.foo()->field = (ctr.foo())->field;
+
+ctr->field->bar = (ctr->field)->bar;
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- [Pin projections] still require library level support via a proc macro and a trait solely for
+  [field types].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This proposal is a lot more general than just improving [pin projections]. It not only covers
+pointer-like types, but also permits all sorts of operations generic over fields.
+
+Not adding this feature will result in the proliferation of `*mut T` over more suitable pointer
+types that better express the invariants of the pointer. The ergonomic cost of
+`unsafe {  MyPtr::new_unchecked(&raw mut (*my_ptr.as_ptr()).field) }` is just to great to be useful
+in practice.
+
+While [pin projections] can be addressed via a library or a separate feature, not having them in the
+language takes a toll on projects trying to minimize dependencies. The Rust for Linux project is
+already using pinning extensively, since all locking primitives require it; a library solution will
+never be as ergonomic as a language-level construct. Thus that project would benefit greatly from
+this feature.
+
+Additionally, safe RCU abstractions are likely impossible without field projections, since they
+require being generic over the fields of structs.
+
+Field projections are on first contact rather difficult to understand, especially the instantiation
+as [pin projections]. However, they are a very natural operation, extending the already existent
+features of raw pointers and references. Therefore they are fairly easy to adjust to; and in turn,
+they provide a big increase in readability of the code, expressing the concept of field projection
+concisely. The compiler changes are rather manageable, reusing several already existing systems,
+thus increasing the maintenance burden only slightly if at all.
+
```

Comment:
```suggestion
We could consider other operators rather than `->`. `->` has associations in C++ with performing a dereference, while field projection doesn't necessarily perform a dereference. For instance, we could consider `~` instead.

```

Reply:
@joshtriplett added your suggestion with some modifications, if you're happy, feel free to resolve this conversation.

Reply:
@y86-dev Looks good to me, thank you.